ccl/testccl/sqlccl: TestTenantTempTableCleanup failed

ccl/testccl/sqlccl.TestTenantTempTableCleanup [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3427750&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3427750&tab=artifacts#/) on master @ [941011c4e582807b40dd03bbcbb8d05385c0638d](https://github.com/cockroachdb/cockroach/commits/941011c4e582807b40dd03bbcbb8d05385c0638d):


```
I210911 05:44:48.256012 14267 sql/table.go:162  [sql,client=127.0.0.1:57056,hostssl,user=root,intExec=delete-temp-TABLE] 412  queued new schema-change job 692285237677096961 for table 53, mutation 0
I210911 05:44:48.326155 12692 5@util/log/event_log.go:32  [sql,client=127.0.0.1:57056,hostssl,user=root] 413 ={"Timestamp":1631339088168176484,"EventType":"drop_table","Statement":"DROP TABLE defaultdb.pg_temp_1631339079503940037_1.temp_table CASCADE","Tag":"DROP TABLE","User":"root","DescriptorID":53,"ApplicationName":"$ internal-delete-temp-TABLE","TableName":"defaultdb.pg_temp_1631339079503940037_1.temp_table"}
I210911 05:44:48.425769 9808 sql/sqlliveness/slinstance/slinstance.go:293  [-] 414  starting SQL liveness instance
I210911 05:44:48.457102 14466 sql/sqlliveness/slstorage/slstorage.go:393  [-] 415  inserted sqlliveness session 01001e8a44a64f43bee742b915685219
I210911 05:44:48.457454 14466 sql/sqlliveness/slinstance/slinstance.go:180  [-] 416  created new SQL liveness session 01001e8a44a64f43bee742b915685219
I210911 05:44:48.481302 9808 1@sql/sqlinstance/instanceprovider/instanceprovider.go:80  [-] 417  created SQL instance 2
I210911 05:44:48.482075 14431 sql/temporary_schema.go:508  [-] 418  running temporary object cleanup background job
I210911 05:44:48.482584 14516 sql/sqlstats/persistedsqlstats/provider.go:142  [-] 419  starting sql-stats-worker with initial delay: 1h1m31.827066879s
I210911 05:44:48.501099 14431 sql/temporary_schema.go:552  [-] 420  found 0 temporary schemas
I210911 05:44:48.501414 14431 sql/temporary_schema.go:555  [-] 421  early exiting temporary schema cleaner as no temporary schemas were found
I210911 05:44:48.501719 14431 sql/temporary_schema.go:556  [-] 422  completed temporary object cleanup job
I210911 05:44:48.501985 14431 sql/temporary_schema.go:639  [-] 423  temporary object cleaner next scheduled to run at 2021-09-11 05:44:49.481911459 +0000 UTC
I210911 05:44:48.569986 12214 2@util/log/event_log.go:32  [-] 424 ={"Timestamp":1631339088569929442,"EventType":"runtime_stats","MemRSSBytes":1000050688,"GoroutineCount":941,"MemStackSysBytes":8847360,"GoAllocBytes":84139760,"GoTotalBytes":155671592,"GoStatsStaleness":0.003296,"HeapFragmentBytes":20029712,"HeapReservedBytes":25919488,"HeapReleasedBytes":62390272,"CGoAllocBytes":7187680,"CGoTotalBytes":15687680,"CGoCallRate":0.0000014950907,"CPUUserPercent":0.0000045643487,"CPUSysPercent":0.0000004848777,"GCPausePercent":0.0000000010288307,"GCRunCount":80,"NetHostRecvBytes":1539681,"NetHostSendBytes":1539681}
I210911 05:44:48.580929 14351 server/settingswatcher/settings_watcher.go:112  [rangefeed=settings] 425  set cluster version to: 21.1-1162
I210911 05:44:48.591823 9808 server/server_sql.go:1089  [-] 426  done ensuring all necessary startup migrations have run
I210911 05:44:48.592423 9808 1@server/tenant.go:272  [config] 427  clusterID: d751c0fe-110c-4e8c-b6f7-81b0d32c494d
I210911 05:44:48.592513 14524 jobs/job_scheduler.go:351  [-] 428  waiting 2m0s before scheduled jobs daemon start
I210911 05:44:48.592782 9808 1@server/tenant.go:273  [config] 429  tenantID: 10
I210911 05:44:48.593129 9808 1@server/tenant.go:273  [config] 430  instanceID: 2
I210911 05:44:48.593355 9808 1@server/server.go:2330  [-] 431  serving sql connections
I210911 05:44:48.647414 12111 sql/temporary_schema.go:508  [-] 432  running temporary object cleanup background job
I210911 05:44:48.665444 11768 sql/temporary_schema.go:500  [n3] 433  skipping temporary object cleanup run as it is not the leaseholder
I210911 05:44:48.665967 11768 sql/temporary_schema.go:501  [n3] 434  completed temporary object cleanup job
I210911 05:44:48.666293 11768 sql/temporary_schema.go:639  [n3] 435  temporary object cleaner next scheduled to run at 2021-09-11 05:44:49.660456573 +0000 UTC
I210911 05:44:48.684020 12111 sql/temporary_schema.go:552  [-] 436  found 0 temporary schemas
I210911 05:44:48.684425 12111 sql/temporary_schema.go:555  [-] 437  early exiting temporary schema cleaner as no temporary schemas were found
I210911 05:44:48.684746 12111 sql/temporary_schema.go:556  [-] 438  completed temporary object cleanup job
I210911 05:44:48.684991 12111 sql/temporary_schema.go:639  [-] 439  temporary object cleaner next scheduled to run at 2021-09-11 05:44:49.642560884 +0000 UTC
I210911 05:44:48.695436 14568 sql/catalog/lease/lease.go:1125  [-] 440  released orphaned lease: {id:33 version:1 expiration:{Time:{wall:732443000 ext:63766936194 loc:<nil>}}}
I210911 05:44:48.700361 14565 sql/catalog/lease/lease.go:1125  [-] 441  released orphaned lease: {id:4 version:1 expiration:{Time:{wall:245301000 ext:63766936181 loc:<nil>}}}
I210911 05:44:48.705652 14567 sql/catalog/lease/lease.go:1125  [-] 442  released orphaned lease: {id:15 version:1 expiration:{Time:{wall:641635000 ext:63766936169 loc:<nil>}}}
I210911 05:44:48.710740 14569 sql/catalog/lease/lease.go:1125  [-] 443  released orphaned lease: {id:35 version:1 expiration:{Time:{wall:871147000 ext:63766936178 loc:<nil>}}}
I210911 05:44:48.718307 14566 sql/catalog/lease/lease.go:1125  [-] 444  released orphaned lease: {id:12 version:1 expiration:{Time:{wall:186910000 ext:63766936176 loc:<nil>}}}
I210911 05:44:48.741333 14618 sql/catalog/lease/lease.go:1125  [-] 445  released orphaned lease: {id:37 version:1 expiration:{Time:{wall:477958000 ext:63766936193 loc:<nil>}}}
W210911 05:44:48.870214 12343 jobs/registry.go:822  [-] 446  canceling all adopted jobs due to stopper quiescing
W210911 05:44:48.870870 14347 jobs/registry.go:822  [-] 447  canceling all adopted jobs due to stopper quiescing
W210911 05:44:48.871061 12229 sql/sqlliveness/slinstance/slinstance.go:217  [-] 448  exiting heartbeat loop
W210911 05:44:48.875061 14416 ccl/kvccl/kvtenantccl/connector.go:181  [sql,tenant-connector] 449  error consuming GossipSubscription RPC: rpc error: code = Canceled desc = context canceled
W210911 05:44:48.875305 12215 ccl/kvccl/kvtenantccl/connector.go:181  [sql,tenant-connector] 450  error consuming GossipSubscription RPC: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:56104->127.0.0.1:35483: use of closed network connection
W210911 05:44:48.875816 14466 sql/sqlliveness/slinstance/slinstance.go:217  [-] 451  exiting heartbeat loop
I210911 05:44:48.884770 9808 testutils/testcluster/testcluster.go:101  [-] 452  TestCluster quiescing nodes
I210911 05:44:48.885874 10092 kv/kvclient/kvcoord/transport_race.go:106  [-] 453  transport race promotion: ran 64 iterations on up to 1767 requests
W210911 05:44:48.888076 11790 jobs/registry.go:822  [-] 454  canceling all adopted jobs due to stopper quiescing
W210911 05:44:48.888914 11376 jobs/registry.go:822  [-] 455  canceling all adopted jobs due to stopper quiescing
W210911 05:44:48.891578 10336 jobs/registry.go:822  [-] 456  canceling all adopted jobs due to stopper quiescing
W210911 05:44:48.893007 11414 sql/sqlliveness/slinstance/slinstance.go:217  [n2] 457  exiting heartbeat loop
W210911 05:44:48.894049 11766 sql/sqlliveness/slinstance/slinstance.go:217  [n3] 458  exiting heartbeat loop
W210911 05:44:48.895637 10186 sql/sqlliveness/slinstance/slinstance.go:217  [n1] 459  exiting heartbeat loop
    testing.go:1092: race detected during execution of test
--- FAIL: TestTenantTempTableCleanup (13.70s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantTempTableCleanup PKG=./pkg/ccl/testccl/sqlccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantTempTableCleanup.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/testccl/sqlccl.TestTenantTempTableCleanup [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3429536&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3429536&tab=artifacts#/) on master @ [6eb9df1f3d460f904b5254c3d32c745eecbedd3f](https://github.com/cockroachdb/cockroach/commits/6eb9df1f3d460f904b5254c3d32c745eecbedd3f):


```
I210912 05:47:35.429683 14355 sql/table.go:162  [sql,client=127.0.0.1:37316,hostssl,user=root,intExec=delete-temp-TABLE] 403  queued new schema-change job 692568900991778817 for table 53, mutation 0
I210912 05:47:35.506083 12633 5@util/log/event_log.go:32  [sql,client=127.0.0.1:37316,hostssl,user=root] 404 ={"Timestamp":1631425655343424271,"EventType":"drop_table","Statement":"DROP TABLE defaultdb.pg_temp_1631425646940765936_1.temp_table CASCADE","Tag":"DROP TABLE","User":"root","DescriptorID":53,"ApplicationName":"$ internal-delete-temp-TABLE","TableName":"defaultdb.pg_temp_1631425646940765936_1.temp_table"}
I210912 05:47:35.616328 9792 sql/sqlliveness/slinstance/slinstance.go:293  [-] 405  starting SQL liveness instance
I210912 05:47:35.648348 10044 sql/temporary_schema.go:508  [n1] 406  running temporary object cleanup background job
I210912 05:47:35.653069 14468 sql/sqlliveness/slstorage/slstorage.go:393  [-] 407  inserted sqlliveness session 43a676259efe444ab165fb848717232c
I210912 05:47:35.653370 14468 sql/sqlliveness/slinstance/slinstance.go:180  [-] 408  created new SQL liveness session 43a676259efe444ab165fb848717232c
I210912 05:47:35.653609 11385 sql/temporary_schema.go:500  [n2] 409  skipping temporary object cleanup run as it is not the leaseholder
I210912 05:47:35.653979 11385 sql/temporary_schema.go:501  [n2] 410  completed temporary object cleanup job
I210912 05:47:35.654249 11385 sql/temporary_schema.go:639  [n2] 411  temporary object cleaner next scheduled to run at 2021-09-12 05:47:36.652970924 +0000 UTC
I210912 05:47:35.654952 10044 sql/temporary_schema.go:552  [n1] 412  found 0 temporary schemas
I210912 05:47:35.655213 10044 sql/temporary_schema.go:555  [n1] 413  early exiting temporary schema cleaner as no temporary schemas were found
I210912 05:47:35.655393 10044 sql/temporary_schema.go:556  [n1] 414  completed temporary object cleanup job
I210912 05:47:35.655629 10044 sql/temporary_schema.go:639  [n1] 415  temporary object cleaner next scheduled to run at 2021-09-12 05:47:36.647917828 +0000 UTC
I210912 05:47:35.674801 9792 1@sql/sqlinstance/instanceprovider/instanceprovider.go:80  [-] 416  created SQL instance 2
I210912 05:47:35.675539 14531 sql/temporary_schema.go:508  [-] 417  running temporary object cleanup background job
I210912 05:47:35.676061 14536 sql/sqlstats/persistedsqlstats/provider.go:142  [-] 418  starting sql-stats-worker with initial delay: 55m33.197003779s
I210912 05:47:35.692137 14531 sql/temporary_schema.go:552  [-] 419  found 0 temporary schemas
I210912 05:47:35.692356 14531 sql/temporary_schema.go:555  [-] 420  early exiting temporary schema cleaner as no temporary schemas were found
I210912 05:47:35.692550 14531 sql/temporary_schema.go:556  [-] 421  completed temporary object cleanup job
I210912 05:47:35.692661 14531 sql/temporary_schema.go:639  [-] 422  temporary object cleaner next scheduled to run at 2021-09-12 05:47:36.675400178 +0000 UTC
I210912 05:47:35.741049 14554 server/settingswatcher/settings_watcher.go:112  [rangefeed=settings] 423  set cluster version to: 21.1-1162
I210912 05:47:35.750612 9792 server/server_sql.go:1089  [-] 424  done ensuring all necessary startup migrations have run
I210912 05:47:35.751264 9792 1@server/tenant.go:272  [config] 425  clusterID: 7d54f79d-0035-447a-a9fe-ec684efe1aca
I210912 05:47:35.751614 9792 1@server/tenant.go:273  [config] 426  tenantID: 10
I210912 05:47:35.751838 9792 1@server/tenant.go:273  [config] 427  instanceID: 2
I210912 05:47:35.751998 14458 jobs/job_scheduler.go:351  [-] 428  waiting 2m0s before scheduled jobs daemon start
I210912 05:47:35.752081 9792 1@server/server.go:2330  [-] 429  serving sql connections
I210912 05:47:35.857724 14610 sql/catalog/lease/lease.go:1125  [-] 430  released orphaned lease: {id:12 version:1 expiration:{Time:{wall:716512000 ext:63767022757 loc:<nil>}}}
I210912 05:47:35.866170 14611 sql/catalog/lease/lease.go:1125  [-] 431  released orphaned lease: {id:15 version:1 expiration:{Time:{wall:388766000 ext:63767022735 loc:<nil>}}}
I210912 05:47:35.868393 14612 sql/catalog/lease/lease.go:1125  [-] 432  released orphaned lease: {id:33 version:1 expiration:{Time:{wall:993892000 ext:63767022743 loc:<nil>}}}
I210912 05:47:35.869536 14613 sql/catalog/lease/lease.go:1125  [-] 433  released orphaned lease: {id:35 version:1 expiration:{Time:{wall:404435000 ext:63767022736 loc:<nil>}}}
I210912 05:47:35.871488 14497 sql/catalog/lease/lease.go:1125  [-] 434  released orphaned lease: {id:4 version:1 expiration:{Time:{wall:687979000 ext:63767022750 loc:<nil>}}}
I210912 05:47:35.896905 14518 sql/catalog/lease/lease.go:1125  [-] 435  released orphaned lease: {id:37 version:1 expiration:{Time:{wall:77038000 ext:63767022753 loc:<nil>}}}
W210912 05:47:36.025067 14550 jobs/registry.go:822  [-] 436  canceling all adopted jobs due to stopper quiescing
W210912 05:47:36.028506 12323 jobs/registry.go:822  [-] 437  canceling all adopted jobs due to stopper quiescing
W210912 05:47:36.028842 14468 sql/sqlliveness/slinstance/slinstance.go:217  [-] 438  exiting heartbeat loop
W210912 05:47:36.029809 12215 ccl/kvccl/kvtenantccl/connector.go:181  [sql,tenant-connector] 439  error consuming GossipSubscription RPC: rpc error: code = Canceled desc = context canceled
W210912 05:47:36.030559 14424 ccl/kvccl/kvtenantccl/connector.go:181  [sql,tenant-connector] 440  error consuming GossipSubscription RPC: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:44424->127.0.0.1:41637: use of closed network connection
W210912 05:47:36.031362 12203 sql/sqlliveness/slinstance/slinstance.go:217  [-] 441  exiting heartbeat loop
I210912 05:47:36.040132 9792 testutils/testcluster/testcluster.go:101  [-] 442  TestCluster quiescing nodes
I210912 05:47:36.041103 10143 kv/kvclient/kvcoord/transport_race.go:106  [-] 443  transport race promotion: ran 72 iterations on up to 1769 requests
W210912 05:47:36.042648 11820 jobs/registry.go:822  [-] 444  canceling all adopted jobs due to stopper quiescing
W210912 05:47:36.045222 11463 jobs/registry.go:822  [-] 445  canceling all adopted jobs due to stopper quiescing
W210912 05:47:36.045371 10042 sql/sqlliveness/slinstance/slinstance.go:217  [n1] 446  exiting heartbeat loop
W210912 05:47:36.048282 10375 jobs/registry.go:822  [-] 447  canceling all adopted jobs due to stopper quiescing
W210912 05:47:36.055121 11781 sql/sqlliveness/slinstance/slinstance.go:217  [n3] 448  exiting heartbeat loop
W210912 05:47:36.055688 11383 sql/sqlliveness/slinstance/slinstance.go:217  [n2] 449  exiting heartbeat loop
W210912 05:47:36.059941 11863 ts/db.go:194  [n3,ts-poll] 450  error writing time series data: failed to send RPC: sending to all replicas failed; last error: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:44392->127.0.0.1:41637: use of closed network connection
    testing.go:1092: race detected during execution of test
--- FAIL: TestTenantTempTableCleanup (13.22s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantTempTableCleanup PKG=./pkg/ccl/testccl/sqlccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantTempTableCleanup.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/testccl/sqlccl.TestTenantTempTableCleanup [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3431531&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3431531&tab=artifacts#/) on master @ [394e35cbbb7ae466383c3ee6f2bbcc76857afeed](https://github.com/cockroachdb/cockroach/commits/394e35cbbb7ae466383c3ee6f2bbcc76857afeed):


```
I210913 05:30:44.117583 14405 sql/table.go:162  [sql,client=127.0.0.1:60208,hostssl,user=root,intExec=delete-temp-TABLE] 411  queued new schema-change job 692848702644256769 for table 53, mutation 0
I210913 05:30:44.179551 12629 5@util/log/event_log.go:32  [sql,client=127.0.0.1:60208,hostssl,user=root] 412 ={"Timestamp":1631511044040890097,"EventType":"drop_table","Statement":"DROP TABLE defaultdb.pg_temp_1631511035308837733_1.temp_table CASCADE","Tag":"DROP TABLE","User":"root","DescriptorID":53,"ApplicationName":"$ internal-delete-temp-TABLE","TableName":"defaultdb.pg_temp_1631511035308837733_1.temp_table"}
I210913 05:30:44.295180 9832 sql/sqlliveness/slinstance/slinstance.go:293  [-] 413  starting SQL liveness instance
I210913 05:30:44.324099 14484 sql/sqlliveness/slstorage/slstorage.go:393  [-] 414  inserted sqlliveness session 7161c98ed3b044e6b2b78327a2e53dd3
I210913 05:30:44.324359 14484 sql/sqlliveness/slinstance/slinstance.go:180  [-] 415  created new SQL liveness session 7161c98ed3b044e6b2b78327a2e53dd3
I210913 05:30:44.347076 9832 1@sql/sqlinstance/instanceprovider/instanceprovider.go:80  [-] 416  created SQL instance 2
I210913 05:30:44.347641 14548 sql/temporary_schema.go:508  [-] 417  running temporary object cleanup background job
I210913 05:30:44.348184 14553 sql/sqlstats/persistedsqlstats/provider.go:142  [-] 418  starting sql-stats-worker with initial delay: 1h6m49.898321231s
I210913 05:30:44.365165 14548 sql/temporary_schema.go:552  [-] 419  found 0 temporary schemas
I210913 05:30:44.365365 14548 sql/temporary_schema.go:555  [-] 420  early exiting temporary schema cleaner as no temporary schemas were found
I210913 05:30:44.365532 14548 sql/temporary_schema.go:556  [-] 421  completed temporary object cleanup job
I210913 05:30:44.365681 14548 sql/temporary_schema.go:639  [-] 422  temporary object cleaner next scheduled to run at 2021-09-13 05:30:45.347487078 +0000 UTC
I210913 05:30:44.415370 14571 server/settingswatcher/settings_watcher.go:112  [rangefeed=settings] 423  set cluster version to: 21.1-1162
I210913 05:30:44.424618 9832 server/server_sql.go:1089  [-] 424  done ensuring all necessary startup migrations have run
I210913 05:30:44.425259 9832 1@server/tenant.go:272  [config] 425  clusterID: e406b192-3680-464b-a8cc-47b01156486f
I210913 05:30:44.425371 14535 jobs/job_scheduler.go:351  [-] 426  waiting 4m0s before scheduled jobs daemon start
I210913 05:30:44.425546 9832 1@server/tenant.go:273  [config] 427  tenantID: 10
I210913 05:30:44.425887 9832 1@server/tenant.go:273  [config] 428  instanceID: 2
I210913 05:30:44.426059 9832 1@server/server.go:2330  [-] 429  serving sql connections
I210913 05:30:44.494350 12199 2@util/log/event_log.go:32  [-] 430 ={"Timestamp":1631511044494310691,"EventType":"runtime_stats","MemRSSBytes":979599360,"GoroutineCount":982,"MemStackSysBytes":10747904,"GoAllocBytes":97788016,"GoTotalBytes":154500136,"GoStatsStaleness":0.003023273,"HeapFragmentBytes":12484496,"HeapReservedBytes":17006592,"HeapReleasedBytes":63299584,"CGoAllocBytes":7130336,"CGoTotalBytes":17326080,"CGoCallRate":0.0000014949331,"CPUUserPercent":0.000004421668,"CPUSysPercent":0.00000045847068,"GCPausePercent":0.00000000089932106,"GCRunCount":79,"NetHostRecvBytes":1640606,"NetHostSendBytes":1640606}
I210913 05:30:44.529601 14635 sql/catalog/lease/lease.go:1125  [-] 431  released orphaned lease: {id:35 version:1 expiration:{Time:{wall:962922000 ext:63767108143 loc:<nil>}}}
I210913 05:30:44.532622 14632 sql/catalog/lease/lease.go:1125  [-] 432  released orphaned lease: {id:12 version:1 expiration:{Time:{wall:392026000 ext:63767108141 loc:<nil>}}}
I210913 05:30:44.536628 14631 sql/catalog/lease/lease.go:1125  [-] 433  released orphaned lease: {id:4 version:1 expiration:{Time:{wall:161822000 ext:63767108147 loc:<nil>}}}
I210913 05:30:44.537325 14633 sql/catalog/lease/lease.go:1125  [-] 434  released orphaned lease: {id:15 version:1 expiration:{Time:{wall:239636000 ext:63767108128 loc:<nil>}}}
I210913 05:30:44.537354 14634 sql/catalog/lease/lease.go:1125  [-] 435  released orphaned lease: {id:33 version:1 expiration:{Time:{wall:304613000 ext:63767108138 loc:<nil>}}}
I210913 05:30:44.557714 11807 sql/temporary_schema.go:508  [-] 436  running temporary object cleanup background job
I210913 05:30:44.579343 11807 sql/temporary_schema.go:552  [-] 437  found 0 temporary schemas
I210913 05:30:44.579573 11807 sql/temporary_schema.go:555  [-] 438  early exiting temporary schema cleaner as no temporary schemas were found
I210913 05:30:44.579769 11807 sql/temporary_schema.go:556  [-] 439  completed temporary object cleanup job
I210913 05:30:44.579922 11807 sql/temporary_schema.go:639  [-] 440  temporary object cleaner next scheduled to run at 2021-09-13 05:30:45.557510825 +0000 UTC
I210913 05:30:44.591973 14543 sql/catalog/lease/lease.go:1125  [-] 441  released orphaned lease: {id:37 version:1 expiration:{Time:{wall:716808000 ext:63767108143 loc:<nil>}}}
I210913 05:30:44.624689 11727 sql/temporary_schema.go:500  [n3] 442  skipping temporary object cleanup run as it is not the leaseholder
I210913 05:30:44.625225 11727 sql/temporary_schema.go:501  [n3] 443  completed temporary object cleanup job
I210913 05:30:44.625605 11727 sql/temporary_schema.go:639  [n3] 444  temporary object cleaner next scheduled to run at 2021-09-13 05:30:45.624470184 +0000 UTC
W210913 05:30:44.723957 14567 jobs/registry.go:822  [-] 445  canceling all adopted jobs due to stopper quiescing
W210913 05:30:44.724818 14484 sql/sqlliveness/slinstance/slinstance.go:217  [-] 446  exiting heartbeat loop
W210913 05:30:44.726076 12330 jobs/registry.go:822  [-] 447  canceling all adopted jobs due to stopper quiescing
W210913 05:30:44.726349 12212 sql/sqlliveness/slinstance/slinstance.go:217  [-] 448  exiting heartbeat loop
W210913 05:30:44.727020 14482 ccl/kvccl/kvtenantccl/connector.go:181  [sql,tenant-connector] 449  error consuming GossipSubscription RPC: rpc error: code = Canceled desc = context canceled
W210913 05:30:44.728010 12200 ccl/kvccl/kvtenantccl/connector.go:181  [sql,tenant-connector] 450  error consuming GossipSubscription RPC: rpc error: code = Canceled desc = context canceled
I210913 05:30:44.736530 9832 testutils/testcluster/testcluster.go:101  [-] 451  TestCluster quiescing nodes
W210913 05:30:44.737960 10062 sql/sqlliveness/slinstance/slinstance.go:217  [n1] 452  exiting heartbeat loop
W210913 05:30:44.739465 11834 jobs/registry.go:822  [-] 453  canceling all adopted jobs due to stopper quiescing
W210913 05:30:44.739873 11442 jobs/registry.go:822  [-] 454  canceling all adopted jobs due to stopper quiescing
W210913 05:30:44.743904 10241 jobs/registry.go:822  [-] 455  canceling all adopted jobs due to stopper quiescing
W210913 05:30:44.747728 11725 sql/sqlliveness/slinstance/slinstance.go:217  [n3] 456  exiting heartbeat loop
W210913 05:30:44.748357 11215 sql/sqlliveness/slinstance/slinstance.go:217  [n2] 457  exiting heartbeat loop
I210913 05:30:44.794904 10181 kv/kvclient/kvcoord/transport_race.go:106  [-] 458  transport race promotion: ran 71 iterations on up to 1773 requests
    testing.go:1092: race detected during execution of test
--- FAIL: TestTenantTempTableCleanup (13.52s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantTempTableCleanup PKG=./pkg/ccl/testccl/sqlccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-race -parallel=4
</p>
</details>
/cc @cockroachdb/sql-queries fqazi
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantTempTableCleanup.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/txn_coord_sender.go
+++ b/pkg/kv/kvclient/kvcoord/txn_coord_sender.go
@@ -1081,6 +1081,8 @@ func (tc *TxnCoordSender) IsSerializablePushAndRefreshNotPossible() bool {
 
 // Epoch is part of the client.TxnSender interface.
 func (tc *TxnCoordSender) Epoch() enginepb.TxnEpoch {
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
 	return tc.mu.txn.Epoch
 }
 

--- a/pkg/kv/txn.go
+++ b/pkg/kv/txn.go
@@ -214,20 +214,31 @@ func (txn *Txn) Epoch() enginepb.TxnEpoch {
 	return txn.mu.sender.Epoch()
 }
 
-// status returns the txn proto status field.
-func (txn *Txn) status() roachpb.TransactionStatus {
+// statusLocked returns the txn proto status field.
+func (txn *Txn) statusLocked() roachpb.TransactionStatus {
 	return txn.mu.sender.TxnStatus()
 }
 
 // IsCommitted returns true iff the transaction has the committed status.
 func (txn *Txn) IsCommitted() bool {
-	return txn.status() == roachpb.COMMITTED
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+	return txn.statusLocked() == roachpb.COMMITTED
+}
+
+// IsAborted returns true iff the transaction has the aborted status.
+func (txn *Txn) IsAborted() bool {
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+	return txn.statusLocked() == roachpb.ABORTED
 }
 
 // IsOpen returns true iff the transaction is in the open state where
 // it can accept further commands.
 func (txn *Txn) IsOpen() bool {
-	return txn.status() == roachpb.PENDING
+	txn.mu.Lock()
+	defer txn.mu.Unlock()
+	return txn.statusLocked() == roachpb.PENDING
 }
 
 // SetUserPriority sets the transaction's user priority. Transactions default to
@@ -670,7 +681,7 @@ func (txn *Txn) CleanupOnError(ctx context.Context, err error) {
 		panic(errors.WithContextTags(errors.AssertionFailedf("CleanupOnError() called with nil error"), ctx))
 	}
 	if replyErr := txn.rollback(ctx); replyErr != nil {
-		if _, ok := replyErr.GetDetail().(*roachpb.TransactionStatusError); ok || txn.status() == roachpb.ABORTED {
+		if _, ok := replyErr.GetDetail().(*roachpb.TransactionStatusError); ok || txn.IsAborted() {
 			log.Eventf(ctx, "failure aborting transaction: %s; abort caused by: %s", replyErr, err)
 		} else {
 			log.Warningf(ctx, "failure aborting transaction: %s; abort caused by: %s", replyErr, err)
@@ -881,7 +892,7 @@ func (txn *Txn) exec(ctx context.Context, fn func(context.Context, *Txn) error) 
 		// Commit on success, unless the txn has already been committed by the
 		// closure. We allow that, as closure might want to run 1PC transactions.
 		if err == nil {
-			if txn.status() != roachpb.COMMITTED {
+			if !txn.IsCommitted() {
 				err = txn.Commit(ctx)
 				log.Eventf(ctx, "client.Txn did AutoCommit. err: %v\n", err)
 				if err != nil {

